python/refactor/main.py

```diff
@@ -0,0 +1,208 @@
+
+def create_dataset_graph(args):
+    g = None
+    dataset = get_dataset(data_path=args.data_path,
+                          data_name=args.dataset,
+                          format_str=args.format,
+                          delimiter=args.delimiter,
+                          files=args.data_files,
+                          has_edge_importance=args.has_edge_importance)
+    train_dataset, eval_dataset, test_dataset = None, None, None
+    # create training dataset needed parameters
+    train_dataset = TrainDataset(dataset, args)
+    g = train_dataset.g
+    args.strict_rel_part = args.mix_cpu_gpu and (train_dataset.cross_part is False)
+    args.soft_rel_part = args.mix_cpu_gpu and args.rel_part and train_dataset.cross_part is True
+    args.rel_parts = train_dataset.rel_parts if args.strict_rel_part or args.soft_rel_part else None
+    args.n_entities = dataset.n_entities
+    args.n_relations = dataset.n_relations
+    args.cross_rels = train_dataset.cross_rels if args.soft_rel_part else None
+
+    if args.neg_sample_size_eval < 0:
+        args.neg_sample_size_eval = dataset.n_entities
+
+    if args.valid or args.test:
+        if args.valid:
+            assert dataset.valid is not None, 'validation set is not provided'
+            eval_dataset = ValidDataset(dataset, args)
+            g = eval_dataset.g
+        if args.test:
+            assert dataset.test is not None, 'test set is not provided'
+            # no sampler is needed
+            test_dataset = TestDataset(dataset, args)
+            g = test_dataset.g
+    return [train_dataset, eval_dataset, test_dataset], g
+
+def create_dataloader(args):
+    train_dataloader, eval_dataloader = None, None
+    if args.model == 'KGE':
+        train_metadata = {'neg_sample_size': args.neg_sample_size,
+                          'chunk_size': args.neg_sample_size,
+                          'chunk': True}
+        train_dataloader = KGETrainDataLoader(shuffle=args.shuffle_data,
+                                              batch_size=args.batch_size,
+                                              pin_memory=args.pin_memory,
+                                              num_workers=args.num_workers,
+                                              drop_last=args.drop_last,
+                                              metadata=train_metadata)
+        eval_metadata = {'num_nodes': args.n_entities,
+                         'chunk_size': 1,
+                         'neg_sample_size': args.n_entities,
+                         'chunk': True}
+        eval_dataloader = KGEEvalDataLoader(batch_size=args.batch_size_eval,
+                                            pin_memory=args.pin_memory,
+                                            num_workers=args.num_workers,
+                                            metadata=eval_metadata)
+    else:
+        raise NotImplementedError(f'Dataloader for {args.model} is not supported yet.')
+    return train_dataloader, eval_dataloader
```

Comment:
Why there is no test_dataloader?

Reply:
Both validation and test follow the same logic that it uses same number of process, validation batch size, negative sample size, etc. while train might take different parameters settings such as chunk training, etc. During the process, the eval_dataloader can take either validation or test dataset and generates corresponding sub data sampler for each process.